bigint/src/lib.rs

```diff
@@ -1349,6 +1351,40 @@ impl Integer for BigUint {
     }
 }
 
+fn high_bits_to_u64(v: &BigUint) -> u64 {
+    match v.data.len() {
+        0   => 0,
+        1   => v.data[0] as u64,
+        _   => {
+            let mut bits = v.bits();
+            let mut ret = 0u64;
+            let mut ret_bits = 0;
+
+            for d in v.data.iter().rev() {
+                let digit_bits = ((bits - 1) & (big_digit::BITS - 1)) + 1;
+                let bits_want = cmp::min(64 - ret_bits, digit_bits);
+
+                if bits_want != 64 {
+                    ret <<= bits_want;
+                }
+                ret      |= *d as u64 >> (digit_bits - bits_want);
+                ret_bits += bits_want;
+                bits     -= bits_want;
+
+                if ret_bits == 64 {
+                    break;
+                }
+            }
+
+            ret
+        }
+    }
+}
+
+fn ilog2(v: u64) -> usize {
+    (64 - v.leading_zeros()) as usize
+}
```

Comment:
This `ilog2` isn't really mathematically correct, e.g. for 4 it will return (64-61)=3.  So it's more like `floor(log2(v))+1`, or in programming terms it's simply the number of significant bits.  Let's rename it to avoid confusion, please.